pkg/controllers/management/auth/globalroles/globalrolebinding_handler.go

```diff
@@ -585,6 +598,127 @@ func (grb *globalRoleBindingLifecycle) grantRestrictedAdminUserClusterPermission
+			continue
+		} else if err != nil {
+			returnError = multierror.Append(returnError, errors.Wrapf(err, "couldn't get namespace %s", ns))
+			continue
+		}
+
+		if namespace.Status.Phase == corev1.NamespaceTerminating {
+			// If the namespace is terminating, don't try to make or update RoleBindings
+			logrus.Warnf("[%v] Namespace %s is terminating. Not creating roles for %s", grController, ns, globalRoleBinding.Name)
+			continue
+		}
+
+		roleName := wrangler.SafeConcatName(gr.Name, ns)
+		roleRef := v1.RoleRef{
+			APIGroup: rbacv1.GroupName,
+			Kind:     "Role",
+			Name:     roleName,
+		}
+
+		subject := rbac.GetGRBSubject(globalRoleBinding)
+		rbName := wrangler.SafeConcatName(globalRoleBinding.Name, ns)
+		roleBinding, err := grb.roleBindingLister.Get(ns, rbName)
+		if err == nil {
+			if reflect.DeepEqual(roleRef, roleBinding.RoleRef) && roleBinding.Labels != nil && roleBinding.Labels[grbOwnerLabel] == grbName {
+				roleBindingUIDs[roleBinding.UID] = struct{}{}
+				continue
+			}
+			// Since roleRef is immutable, we have to delete and recreate the RB
+			err = grb.roleBindings.DeleteNamespaced(roleBinding.Namespace, roleBinding.Name, &metav1.DeleteOptions{})
+			if err != nil {
+				returnError = multierror.Append(returnError, err)
+				continue
+			}
+		} else if !apierrors.IsNotFound(err) {
+			returnError = multierror.Append(returnError, err)
+			continue
+		}
+
+		// Create a new RoleBinding
+		newRoleBinding := &v1.RoleBinding{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      rbName,
+				Namespace: ns,
+				OwnerReferences: []metav1.OwnerReference{
+					{
+						APIVersion: globalRoleBinding.APIVersion,
+						Kind:       globalRoleBinding.Kind,
+						UID:        globalRoleBinding.UID,
+						Name:       globalRoleBinding.Name,
+					},
+				},
+				Labels: map[string]string{
+					grbOwnerLabel: grbName,
+				},
+			},
+			RoleRef:  roleRef,
+			Subjects: []v1.Subject{subject},
+		}
+
+		createdRB, err := grb.roleBindings.Create(newRoleBinding)
+		if err != nil {
+			returnError = multierror.Append(returnError, err)
+			continue
+		}
+		roleBindingUIDs[createdRB.UID] = struct{}{}
+	}
+
+	// get all the roleBindings claiming to be owned by this GRB and remove any that shouldn't exist
+	r, err := labels.NewRequirement(grbOwnerLabel, selection.Equals, []string{grbName})
+	if err != nil {
+		return multierror.Append(returnError, errors.Wrapf(err, "couldn't create label: %s", grOwnerLabel))
+	}
+
+	rbs, err := grb.roleBindingLister.List("", labels.NewSelector().Add(*r))
+	if err != nil {
+		return multierror.Append(returnError,
+			errors.Wrapf(err, "couldn't list roleBindings with label %s : %s", grbOwnerLabel, grbName))
+	}
+
+	// After creating/updating all RBs, if the number of RBs with the grbOwnerLabel is the same as
+	// as the number of created/updated RBs, we know there are no invalid RBs to purge
+	if len(rbs) != len(roleBindingUIDs) {
```

Comment:
Same comment here about the dangers of not purging based on the number of active items.